src/aoai-api-simulator/src/aoai_api_simulator/main.py

```diff
@@ -1,33 +1,91 @@
+from opentelemetry.sdk._logs import LoggerProvider, LoggingHandler
+from opentelemetry.sdk._logs.export import BatchLogRecordProcessor
+from opentelemetry.sdk.metrics import MeterProvider
+from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
+from opentelemetry.sdk.resources import Resource
+from opentelemetry.sdk.trace import TracerProvider
+from opentelemetry.sdk.trace.export import BatchSpanProcessor
 
 log_level = os.getenv("LOG_LEVEL") or "INFO"
 
 logger = logging.getLogger(__name__)
 logging.basicConfig(level=log_level)
 logging.getLogger("azure").setLevel(logging.WARNING)
 
+opentelemetry_exporter_otlp_endpoint = os.getenv("OTEL_EXPORTER_OTLP_ENDPOINT")
 application_insights_connection_string = os.getenv("APPLICATIONINSIGHTS_CONNECTION_STRING")
+
+using_azure_monitor: bool
+
 if application_insights_connection_string:
     logger.info("🚀 Configuring Azure Monitor telemetry")
 
     # Options: https://github.com/Azure/azure-sdk-for-python/tree/main/sdk/monitor/azure-monitor-opentelemetry#usage
     configure_azure_monitor(connection_string=application_insights_connection_string)
+    using_azure_monitor = True
 else:
+    using_azure_monitor = False
     logger.info("🚀 Azure Monitor telemetry not configured (set APPLICATIONINSIGHTS_CONNECTION_STRING)")
 
-# tracer = trace.get_tracer(__name__)
+    if opentelemetry_exporter_otlp_endpoint:
+        logger.info("🚀 Configuring OTLP telemetry")
+
+        # setup the instrumentors
+        resource = Resource(attributes={"service.name": os.getenv("OTEL_SERVICE_NAME", "aoai-api-simulator")})
+
+        trace.set_tracer_provider(TracerProvider(resource=resource))
+        tracer = trace.get_tracer(__name__)
+
+        otlp_exporter = OTLPSpanExporter()
+
+        # tracing
+        tracer = trace.get_tracer(__name__)
+        span_processor = BatchSpanProcessor(otlp_exporter)
+        trace.get_tracer_provider().add_span_processor(span_processor)
+
+        # metrics
+        reader = PeriodicExportingMetricReader(OTLPMetricExporter())
+        meterProvider = MeterProvider(resource=resource, metric_readers=[reader])
+        metrics.set_meter_provider(meterProvider)
+
+        # logging
+        logger_provider = LoggerProvider(
+            resource=resource,
+        )
+
+        otlp_exporter = OTLPLogExporter()
+        logger_provider.add_log_record_processor(BatchLogRecordProcessor(otlp_exporter))
+
+        handler = LoggingHandler(level=os.getenv("OTEL_LOG_LEVEL", "INFO"), logger_provider=logger_provider)
+        # Attach OTLP handler to root logger
+        logging.getLogger().addHandler(handler)
+    else:
+        logger.info("🚀 OTLP telemetry exporter not configured (set OTEL_EXPORTER_OTLP_ENDPOINT)")
 
 config = get_config_from_env_vars(logger)
 set_config(config)
 
-
 apply_config()
+
 app = builder_app  # expose to gunicorn
+
+if not using_azure_monitor:
+    RequestsInstrumentor().instrument()
```

Comment:
Are these needed to be able to get the metrics exported via OTLP?

Reply:
Yes but only if the Azure OTEL distro has not already done it. I've made it more clear.

Reply:
So you can only get the simulator custom metrics exported if you also do the request intstrumentation?

My recollection from when I last did a bunch of perf testing/tuning was that the instrumentation caused a noticeable perf hit.

Reply:
Requests instrumentation is automatically enabled (https://learn.microsoft.com/en-us/python/api/overview/azure/monitor-opentelemetry-readme?view=azure-python#officially-supported-instrumentations) with the Azure OpenTelemetry distribution we were using before. This simply adds the same behaviour when using the OTLP exporter option as well. It's not something new we have introduced.

We can however do it only if a OTLP endpoint if specified.